### DISCUSSION TITLE
Should Key have always an expiring date ?

### DISCUSSION THREAD

**Discussion by juchom** [meilisearch/product#403] [Feedback & Feature Proposal]
While I was working on updating a key in the C# client, I faced an issue with current design for expiring date for keys.

https://github.com/meilisearch/meilisearch-dotnet/pull/245#discussion_r814295152

Actually if a key has a date it will expire at this date. If it's null, the key will never expire.

The doc for updating a key says : 
> Updates to keys are partial. This means you should provide only the fields you intend to update, as any fields not present in the payload will remain unchanged.

In strongly typed language JsonSerializer can be configured to remove null values from objects or to keep it. Which can lead to errors for your customers if we try to create a client that sticks to your api

Imagine we have this object for our call to the api

```csharp
public class UpdateKeyInfo
{
    public string Id { get; set; }
    public List<string> Actions { get; set; }
    public DateTime? ExpiresAt { get; set; }
}
```

If the JsonSerializer removes null

If a user creates an instance of this object because he wants to only change the Actions list but not the expiring date, the call will do what the user wants.

If a user wants to have a key that never expires and set null to this object, the value won't be sent.

If the JsonSerializer keeps null

If a user creates an instance of this object because he wants to only change the Actions list but not the expiring date, the call will also set the value null to the expiring date which is not expected.

If a user wants to have a key that never expires and set null to this object and leave null for Actions, the value for the date will be sent but the a null list of actions will be sent also.

I think, it would be better to always have a value for the expiring date. If you want it to expires in one year from now set it to one year from now, if you want it to never expires set the value to 9999-12-31T23:59:59.9999999 or something very far in the future.

In Azure Portal they manage keys like that, a key that never expires is set to expires in a century or something like that.

Doing so, would make it predicable for users. This will follow the same rule for every fields : If it's null don't send it, if it's not send it.

But in any case, you have to set a value when you create the key or if you want to change the expiry date.

**Comment by gmourier** [comment]
Hi @juchom 👋

Thanks for this feedback.

> I think, it would be better to always have a value for the expiring date. If you want it to expires in one year from now set it to one year from now, if you want it to never expires set the value to 9999-12-31T23:59:59.9999999 or something very far in the future.

IMO, it can be counter-intuitive regarding the usage of an API. It forces users to set a value while they don't necessarily want to. It adds another step during their workflow while a configured serializer seems to solve the need to match the behavior of an API using the `null` marker to define a property. 🤔

We thought about forcing an `expiresAt` for security reasons but decided not to go in that direction as it may be better for the developer in terms of experience when testing and configuring Meilisearch during the development cycle. Nevertheless, we have chosen to use the `null` marker so that the user must explicitly define a non-expiring key if he makes this choice.

> Which can lead to errors for your customers if we try to create a client that sticks to your API

Can you elaborate on this? Are you speaking about a front-end client or a custom SDK? 😇

Thanks!

**Comment by juchom** [comment]
It's C# code, if something is not clear do not hesitate to ask

To update a key the doc says this : 

> Updates to keys are partial. This means you should provide only the fields you intend to update, as any fields not present in the payload will remain unchanged.

In a strongly typed language, the contract for updating an keep would be :

```csharp
public class UpdateKey
{
	public string Key { get; set; }
	public string Description { get; set; }
	public List<string> Actions { get; set; }
	public List<string> Indexes { get; set; }
	public DateTime? ExpiresAt { get; set; }
}
```

In our code we can tell the JsonSerializer to remove or keep null values, here is our two configurations

```csharp
var removeNulls = new JsonSerializerOptions
{
	DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
	PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
};

var keepNulls = new JsonSerializerOptions
{
	DefaultIgnoreCondition = JsonIgnoreCondition.Never,
	PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
};
```

Imagine we want to update our key and remove the expiry :

```csharp
var key = new UpdateKey
{
	Key = "d0552b41536279a0ad88bd595327b96f01176a60c2243e906c52ac02375f9bc4",
	Actions = new List<string> { "search", "documents.get" },
	ExpiresAt = null,
};
```

If we serialize this object with removing null values we will have this json

```json
{
  "key": "d0552b41536279a0ad88bd595327b96f01176a60c2243e906c52ac02375f9bc4",
  "actions": [
    "search",
    "documents.get"
  ]
}
```

As you can see the payload will miss the ExpiresAt because it is null.

If we keep null values we will have this json

```json
{
  "key": "d0552b41536279a0ad88bd595327b96f01176a60c2243e906c52ac02375f9bc4",
  "description": null,
  "actions": [
    "search",
    "documents.get"
  ],
  "indexes": null,
  "expiresAt": null
}
```

In this case we will keep all values but we will overwrite `description` and `indexes`.

This means that null values are to be excluded for everything but for the `expiresAt` property the null value has a real meaning and should not be ignored.

I think it's better to always have an expiry date, a user may not always read the documentation and could create a key that never expires which may not be what he wanted or may create a security hole.

Having `expiresAt` mandatory at the creation makes clear for everybody when the key will expire and if you don't want to expire set this date very far in the future.

**Reply by gmourier** [reply]
Thanks @juchom! I understand it clearer now 😇

@meilisearch/integration-team What do you think about this?

Is there any way to have a mechanism on the SDK side that doesn't impact users of an untyped language otherwise and thus doesn't force this change for everyone? I know that @bidoubiwa had pointed out that it was useful for a developer who makes tests not to have to put an expiration date, the problem is that we may have this problem on other fields one day where the value `null` could be used as a marker without being related to a pure security argument.

Would it be bad to allow an empty string instead of null? It doesn't look good to me but I may be wrong.

**Reply by brunoocasali** [reply]
I'm not sure if adding a date in the future is the semantically correct option for this case since we know that the key still has an expiration date.

We could adopt some patterns to prevent problems like this without having to let go of the `null` way. 

#### v1) Ignoring the expiresAt in the PATCH requests.
- Prevent the update of the `expiresAt` field after its creation?
  - Only allow updating the `description`, `indexes`, and `actions ` fields.

I think this could be done quite easily by the core team since they could just ignore the fields from the request silently but this **will affect everyone**.

#### v2) Change the way we handle this in the SDK's

I think this is one of our best options since it looks like a real-world example (talking specifically about the c#):
We need to use the `Constants.JsonSerializerOptionsWriteNulls`.

And instead of creating a new object with the changed fields, we load the object and change the fields we want to.
In this particular case, we will not be using the partial "PATCH" update way of updating, but we will be able to keep things working.

_The main thing is to load the object before the update instead of creating a new instance._
```c#
// suppose we are in an controller handler action:
var myKey = await meilisearchClient.GetKeyAsync(keyUidFromRequest); // myKey will be loaded entirely.
var keyParamsFromRequest = new Key(...);

meilisearchClient.UpdateKeyAsync(myKey.KeyUid, keyParamsFromRequest);
``` 

<details>
  <summary>Click to expand the code sample/test nº1!</summary>
  
  ```c#
    public async Task UpdateOneKey()
        {
            var newKey = new Key
            {
                Description = "Just a new key.",
                Actions = new string[] { ApiActions.DocumentsAdd },
                Indexes = new string[] { "*" },
                ExpiresAt = null
            };
            var createdKey = await _client.CreateKeyAsync(newKey);
            // in a near future, we want to update the key...
            // we retrieve the key
            var myKey = await _client.GetKeyAsync(createdKey.KeyUid);
            // make the necessary changes to it
            myKey.Actions = new string[] { ApiActions.DumpsCreate, ApiActions.IndexesCreate };

            // update the key
            var updatedKey = await _client.UpdateKeyAsync(myKey.KeyUid, myKey);
            // sent json was: 
            // ---------------
            // {"key":"bnj3utD38bf60738cce9a939c39166ab6aada75068a17f5d2c25d6c91c68a13233bfe31a","description":"Just a new key.","actions":["dumps.create","indexes.create"],"indexes":["*"],"expiresAt":null,"createdAt":"2022-04-04T21:11:01.6289767Z","updatedAt":"2022-04-04T21:11:01.6289767Z"}
            // ---------------

            var fetchedKey = await _client.GetKeyAsync(myKey.KeyUid);

            Assert.Equal(fetchedKey.Description, updatedKey.Description);
            Assert.Equal(fetchedKey.KeyUid, updatedKey.KeyUid);
            Assert.Equal(fetchedKey.Indexes, updatedKey.Indexes);
            Assert.Equal(fetchedKey.Actions, new string[] { ApiActions.DumpsCreate, ApiActions.IndexesCreate });
            Assert.Null(fetchedKey.ExpiresAt);
            Assert.Equal(fetchedKey.CreatedAt, updatedKey.CreatedAt);
            Assert.Equal(fetchedKey.UpdatedAt, updatedKey.UpdatedAt);
        }
  ```
</details>

<details>
  <summary>Click to expand the code sample/test nº2!</summary>
  
  ```c#
public async Task UpdateOneKey2()
        {
            var newKey = new Key
            {
                Description = "Just a new key.",
                Actions = new string[] { ApiActions.DocumentsAdd },
                Indexes = new string[] { "*" },
                ExpiresAt = DateTime.Parse("2023-05-05")
            };
            var createdKey = await _client.CreateKeyAsync(newKey);
            // in a near future, we want to update the key...
            // we retrieve the key
            var myKey = await _client.GetKeyAsync(createdKey.KeyUid);
            // make the necessary changes to it
            myKey.Actions = new string[] { ApiActions.DumpsCreate, ApiActions.IndexesCreate };

            // update the key
            var updatedKey = await _client.UpdateKeyAsync(myKey.KeyUid, myKey);
            // sent json was: 
            // ---------------
            // {"key":"OGAx7xTQ11cd13c8a90414838c117ff626b5bad6004b606e12e9b3fa389e2bd41920bae2","description":"Just a new key.","actions":["dumps.create","indexes.create"],"indexes":["*"],"expiresAt":"2023-05-05T00:00:00Z","createdAt":"2022-04-04T21:13:03.8055318Z","updatedAt":"2022-04-04T21:13:03.8055318Z"}
            // ---------------

            var fetchedKey = await _client.GetKeyAsync(myKey.KeyUid);

            Assert.Equal(fetchedKey.Description, updatedKey.Description);
            Assert.Equal(fetchedKey.KeyUid, updatedKey.KeyUid);
            Assert.Equal(fetchedKey.Indexes, updatedKey.Indexes);
            Assert.Equal(fetchedKey.Actions, new string[] { ApiActions.DumpsCreate, ApiActions.IndexesCreate });
            Assert.Equal(fetchedKey.ExpiresAt, DateTime.Parse("2023-05-05"));
            Assert.Equal(fetchedKey.CreatedAt, updatedKey.CreatedAt);
            Assert.Equal(fetchedKey.UpdatedAt, updatedKey.UpdatedAt);
        }
  ```
</details>

<details>
  <summary>Click to expand the code sample/test nº2!</summary>
  
  ```c#
public async Task UpdateOneKey3()
        {
            var newKey = new Key
            {
                Description = "Just a new key.",
                Actions = new string[] { ApiActions.DocumentsAdd },
                Indexes = new string[] { "*" },
                ExpiresAt = DateTime.Parse("2023-05-05")
            };
            var createdKey = await _client.CreateKeyAsync(newKey);
            // in a near future, we want to update the key...
            // we retrieve the key
            var myKey = await _client.GetKeyAsync(createdKey.KeyUid);
            // make the necessary changes to it
            myKey.ExpiresAt = null;

            // update the key
            var updatedKey = await _client.UpdateKeyAsync(myKey.KeyUid, myKey);
            // sent json was: 
            // ---------------
            // {"key":"Zb8KSfp351f51799535335f954a45df84ceb123fdce3a98903eb5c6594ad170205dfef2d","description":"Just a new key.","actions":["documents.add"],"indexes":["*"],"expiresAt":null,"createdAt":"2022-04-04T21:15:36.6361961Z","updatedAt":"2022-04-04T21:15:36.6361961Z"}
            // ---------------

            var fetchedKey = await _client.GetKeyAsync(myKey.KeyUid);

            Assert.Equal(fetchedKey.Description, updatedKey.Description);
            Assert.Equal(fetchedKey.KeyUid, updatedKey.KeyUid);
            Assert.Equal(fetchedKey.Indexes, updatedKey.Indexes);
            Assert.Equal(fetchedKey.Actions, updatedKey.Actions);
            Assert.Null(fetchedKey.ExpiresAt);
            Assert.Equal(fetchedKey.CreatedAt, updatedKey.CreatedAt);
            Assert.Equal(fetchedKey.UpdatedAt, updatedKey.UpdatedAt);
        }
  ```
</details>

All these examples above are passing, with a mixture between your submission @juchom https://github.com/meilisearch/meilisearch-dotnet/pull/245 and the change of keeping json nulls.

@juchom, do you agree with this being the most accurate version? Or should we have a way to handle the partial updates from scratch? Because I'm convinced, this is how most people will use the SDK (loading the key and sending all the parameters they want to change at once).

@gmourier, sorry but I think I didn't answer your question, but I gave a new look to this question which could replace a direct answer for you :)

**Reply by juchom** [reply]
Ok, I'm sure this will not work because of :

> The main thing is to load the object before the update instead of creating a new instance.

You have no guarantee a user will always do this, which will lead to strange behaviours. And you can be sure this will happen !

I think sample 1 and 2 are the same and suppose that you always send the expiration date on update and this is the main issue.

What will happen if the user use directly the key id and create a new KeyObject with only updated values, in our case just the `Actions` ?
- It will silently remove expiry date.
- What will be the behaviour of Meilisearch if it receives a null value for the `Indexes` ?

I really think that the first question you have to answer is do you want to use the null value for expiration date ? I, personally, don't  like it, you give the user the ability to create keys that never expire silently and you are getting in trouble with strongly typed languages because you have to make assumptions on how your users will have to use the sdk in order to respect the way the expiry date works. And we just saw that if the dev are not respecting your assumptions this may lead to silently change the expiration date of the key.

If you want to keep the null value, then it will be safer to split the update of actions and indexes in one function and the expiration date in another. It won't match a one to one api endpoint / c# function but I don't see any other options at the moment.

The first thing is you have to discuss with your team about the meaning of `null` and see if you want to have an exception for the expiration date. Then we will work on making the thing safe for your customers.

This has nothing to do with our issue, but you should consider having per `Key` two `AccessKey`, this will help your customers rotating their `AccessKey` without service discontinuation. If they want to update `AccessKey1`, they update their application to use `AccessKey2`, then they can update `AccessKey1`.

**Reply by brunoocasali** [reply]
> I really think that the first question you have to answer is, do you want to use the null value for the expiration date? 

Well, my main concern is that given the possibility to have the `null` value, users can have keys that never expire. For me, this is the only problem I see as a security concern because semantically, as I said before, it is correct that "if we support a key that never expires", then "we should have no date". 
I think this could be considered somewhere in the future @gmourier (require the user to set a date every time - even if this will hurt the development/testing cycles).

Since we cannot guarantee anything before the method invocation (I agree with you on this juchom). To fix the problem in the C# and possibly in other strongly typed languages we can work with your suggestion @juchom:

> If you want to keep the null value, then it will be safer to split the update of actions and indexes in one function and the expiration date in another. It won't match a one to one api endpoint / c# function but I don't see any other options at the moment.



@gmourier for future references we can discuss some options about this:
  - Remove the possibility of partial updates for Keys.
  - Prevent update expiresAt at Keys, but to do this we probably will need something like this https://github.com/meilisearch/product/discussions/337 to help.
  - Create a new endpoint just to update `expiresAt` like `/keys/:key_hash/remove-expiration`.

**Reply by juchom** [reply]
I was thinking at this problem and given the actual definition of a key, you are right setting a date far in the future to say it's not expiring is wrong.

I think the main issue is to give the possibility to create a key that never expires for something that contains access rights.

Since you want to offer Meilisearch as a hosted service, I think for a cloud service it has to be clear that the expiry date is a mandatory attribute of your keys. Just let your customers decide what they want, they want keys that expires every month or in a century let them decide.

I let you decide if you want to change the definition of your access keys to always expiring, and make the expiry date field mandatory.

Again, since you want to offer Meilisearch as a service, you should definitely consider having to access key per key to allow your customers rotating their access keys without service interruption.

Let me know what you decide and I will update the .net client accordingly.

**Reply by gmourier** [reply]
Hi @juchom 👋 

We had a working group about this with @bidoubiwa  and @brunoocasali today.

We'll make `expiresAt` to have an expiration date defined in any case, we believe it's the most obvious and simplest solution to avoid this problem.

Also, it's more secure by design which is a good thing in the long run. A key that never expires can be problematic and users can always set a date far into the future as a work-around.

This change is planned for v0.28.

Thanks a lot for your help on this subject and for bringing up that issue to the table. ❤️

**Reply by juchom** [reply]
Ok, this is a good news !

While you are working on this topic, even if this is not linked to our issue, will you add 2 access keys per key to help in rotating keys ?

And thanks for making this great product ! 👍

**Reply by gmourier** [reply]
I have created another product discussion so that we can discuss it specifically and make it easier for others to contribute.

https://github.com/meilisearch/product/discussions/438

**Reply by gmourier** [reply]
Hey 👋

A quick update; keys can still have no expiration date if **desired** (as currently, it is necessary to specify `null` at creation).

The change for v0.28 is that the security-related fields (`actions`/ `indexes` and `expiresAt`) can no longer be modified for an API key.

**Comment by gmourier** [comment]
Following the [v0.28 release](https://github.com/meilisearch/meilisearch/releases/tag/v0.28.0), this discussion is now locked.

We encourage you to create a new thread if needed.

Thank you ✨
